### PR TITLE
[Feature] Support PAConv modules

### PR DESCRIPTION
**Pull request by Wuziyi616**
Replace `Conv2d` with `PAConv` and we get set abstraction layers in PAConv's paper. In order to reduce duplicated work, I largely refactor the code under `point_sa_module.py`. I add a base class `BasePointSAModule` so that all other SA layers (in PN++, PAConv or more in the future) can inherit from this base class.

Again, `PAConv` and `PAConvCUDA` need different SA layers. I have tried my best to add comments. Feel free to leave comments if anything could be improved.

PS. The drop of unit test coverage is because we introduce new CUDA op in PAConv.

### CODE REVIEW

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+from torch import nn as nn
+
+from mmdet3d.ops import PAConv, PAConvCUDA
+from .builder import SA_MODULES
+from .point_sa_module import BasePointSAModule
+
+
+@SA_MODULES.register_module()
+class PAConvSAModuleMSG(BasePointSAModule):
+    """Point set abstraction module with multi-scale grouping (MSG) used in
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/paconv_sa_module.py:11]: r"""

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            fps_sample_range_list=fps_sample_range_list,
+            normalize_xyz=normalize_xyz,
+            paconv_kernel_input=paconv_kernel_input,
+            scorenet_input=scorenet_input,
+            scorenet_cfg=scorenet_cfg)
+
+
+@SA_MODULES.register_module()
+class PAConvCUDASAModuleMSG(BasePointSAModule):
+    """Point set abstraction module with multi-scale grouping (MSG) used in
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/paconv_sa_module.py:154]: r"""

File: mmdet3d/ops/pointnet_modules/point_sa_module.py [hunk 3]
```diff
@@ -49,12 +47,12 @@ def __init__(self,
...
                  dilated_group: bool = False,
-                 norm_cfg: dict = dict(type='BN2d'),
                  use_xyz: bool = True,
-                 pool_mod='max',
+                 pool_mod: str = 'max',
                  normalize_xyz: bool = False,
-                 bias='auto'):
-        super().__init__()
+                 grouper_return_grouped_xyz: bool = False,
+                 grouper_return_grouped_idx: bool = False):
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/point_sa_module.py:54]: Reformat these input arguments to be consistent with ours?

File: mmdet3d/ops/pointnet_modules/point_sa_module.py [hunk 4]
```diff
@@ -95,36 +94,78 @@ def __init__(self,
...
                 grouper = GroupAll(use_xyz)
             self.groupers.append(grouper)
 
-            mlp_spec = mlp_channels[i]
-            if use_xyz:
-                mlp_spec[0] += 3
+    def _sample_points(
+            self, points_xyz: torch.Tensor, features: torch.Tensor,
+            indices: torch.Tensor,
+            target_xyz: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/point_sa_module.py:107]: Reformat args

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py (hunk 1 above)
Comment (Tai-Wang) [reply mmdet3d/ops/pointnet_modules/paconv_sa_module.py:11]: Please also fix others

File: mmdet3d/ops/pointnet_modules/point_sa_module.py (hunk 3 above)
Comment (Tai-Wang) [reply mmdet3d/ops/pointnet_modules/point_sa_module.py:54]: Please also reformat others

File: mmdet3d/ops/pointnet_modules/point_sa_module.py [hunk 5]
```diff
@@ -95,36 +94,78 @@ def __init__(self,
...
+        return new_xyz, indices
+
+    def _pool_features(self, features: torch.Tensor) -> torch.Tensor:
+        """Perform feature aggregation using pooling operation.
+
+        Args:
+            features (torch.Tensor): (B, C, N, K)
+
+        Returns:
+            torch.Tensor: (B, C, N)
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/point_sa_module.py:147]: Refine the docstrings

File: mmdet3d/ops/pointnet_modules/point_sa_module.py [hunk 6]
```diff
@@ -144,44 +185,116 @@ def forward(
...
+            fps_sample_range_list=fps_sample_range_list,
+            dilated_group=dilated_group,
+            use_xyz=use_xyz,
+            pool_mod=pool_mod,
+            normalize_xyz=normalize_xyz)
+
+        for i in range(len(self.mlp_channels)):
+            mlp_spec = self.mlp_channels[i]
+            if use_xyz:
+                mlp_spec[0] += 3
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/point_sa_module.py:277]: What does 'spec' mean? Maybe use other expressions?

Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/point_sa_module.py:277]: Oh, the original implementation uses `spec` so I keep it here. See [this](https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/ops/pointnet_modules/point_sa_module.py#L103). I will change it to a meaningful name.

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py (hunk 1 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/paconv_sa_module.py:11]: Fix.

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py (hunk 2 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/paconv_sa_module.py:154]: Fix.

File: mmdet3d/ops/pointnet_modules/point_sa_module.py (hunk 3 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/point_sa_module.py:54]: Done.

File: mmdet3d/ops/pointnet_modules/point_sa_module.py (hunk 4 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/point_sa_module.py:107]: Yes.

File: mmdet3d/ops/pointnet_modules/point_sa_module.py (hunk 5 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/point_sa_module.py:147]: Fix.

File: mmdet3d/ops/pointnet_modules/point_sa_module.py (hunk 6 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/point_sa_module.py]: Can't come up with a better name. Simply change `ml_spec` to `mlp_channel`.

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+from .point_sa_module import BasePointSAModule
+
+
+@SA_MODULES.register_module()
+class PAConvSAModuleMSG(BasePointSAModule):
+    r"""Point set abstraction module with multi-scale grouping (MSG) used in
+    PAConv networks.
+
+    Replace the MLPs in `PointSAModuleMSG` with PAConv layers.
+    See the `paper <https://arxiv.org/abs/2103.14635>`_ for more detials.
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/paconv_sa_module.py:15]: details

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+            self.mlps.append(mlp)
+
+
+@SA_MODULES.register_module()
+class PAConvSAModule(PAConvSAModuleMSG):
+    r"""Point set abstraction module with single-scale grouping (SSG) used in
+    PAConv networks.
+
+    Replace the MLPs in `PointSAModule` with PAConv layers. See the `paper
+    <https://arxiv.org/abs/2103.14635>`_ for more detials.
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/paconv_sa_module.py:113]: details

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+
+@SA_MODULES.register_module()
+class PAConvCUDASAModuleMSG(BasePointSAModule):
+    r"""Point set abstraction module with multi-scale grouping (MSG) used in
+    PAConv networks.
+
+    Replace the non CUDA version PAConv with CUDA implemented PAConv for
+    efficient computation. See the `paper <https://arxiv.org/abs/2103.14635>`_
+    for more detials.
```
Comment (Tai-Wang) [inline mmdet3d/ops/pointnet_modules/paconv_sa_module.py:159]: details

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py (hunk 7 above)
Comment (Tai-Wang) [reply mmdet3d/ops/pointnet_modules/paconv_sa_module.py:15]: Also check other docstrings, seems like there are many same typos

**Review by Tai-Wang — APPROVED** [review]
Just need to fix typos

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py (hunk 7 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/paconv_sa_module.py]: Fix.

File: mmdet3d/ops/pointnet_modules/paconv_sa_module.py (hunk 8 above)
Comment (Wuziyi616) [reply mmdet3d/ops/pointnet_modules/paconv_sa_module.py]: Fix.

**Pull request merged**